### PR TITLE
minmax svgd and function value based par_vi

### PR DESCRIPTION
**Pull request by runjerry**
- add minmax_svgd (Fisher Neural Sampler) to generator par_vi methods.
- add function value based par_vi to hypernetwork algorithm.
- add a simple relu MLP which enables explicit computation diagonals of input-output Jacobian.

### CODE REVIEW

File: alf/networks/relu_mlp.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        inputs = inputs.squeeze()
+        assert inputs.shape[-1] == self._input_size, \
+            ("inputs should has shape {}!".format(self._input_size))
+
+        self.forward(inputs)
+
+        return self._compute_jac_diag()
+
+    def _compute_jac_diag(self):
+        """Compute diagonals of the input-output Jacobian. """
```
Comment (emailweixu) [inline alf/networks/relu_mlp.py:131]: Need reference.

File: alf/algorithms/hypernetwork_algorithm.py [hunk 2]
```diff
@@ -132,6 +136,13 @@ def __init__(self,
             num_particles (int): number of sampling particles
             entropy_regularization (float): weight of entropy regularization
 
+            Args for the critic (used when par_vi is 'minmax')
```
Comment (emailweixu) [inline alf/algorithms/hypernetwork_algorithm.py:139]: Take a look at the compiled documentation to make sure it display correctly.
And see https://alf.readthedocs.io/en/latest/notes/preview_doc_website.html for viewing the doc locally.

File: alf/algorithms/generator.py [hunk 3]
```diff
@@ -30,9 +32,94 @@
                                ["generator", "mi_estimator"])
 
 
+@gin.configurable
+class CriticAlgorithm(Algorithm):
```
Comment (emailweixu) [inline alf/algorithms/generator.py:36]: It is not really an Algorithm. It should still be a Network.

File: alf/algorithms/generator.py [hunk 4]
```diff
@@ -456,11 +634,89 @@ def _gfsf_grad(self, inputs, outputs, loss_func, entropy_regularization):
...
+    def _jacobian_trace(self, fx, x):
+        """Hutchinson's trace Jacobian estimator O(1) call to autograd,
+            used by "\"minmax\" method"""
+        eps = torch.randn_like(fx)
+        jvp = torch.autograd.grad(
+            fx, x, grad_outputs=eps, retain_graph=True, create_graph=True)[0]
+        if eps.shape[-1] == jvp.shape[-1]:
+            tr_jvp = torch.einsum('bi,bi->b', jvp, eps)
+        else:
+            tr_jvp = torch.einsum('bi,bj->b', jvp, eps)
```
Comment (emailweixu) [inline alf/algorithms/generator.py:651]: When can this happen? This doesn't seem to be a trace anymore

**Comment by runjerry** [comment]
included in the new PR #742

**Pull request closed without merging**